### PR TITLE
fix: enable upsert protobuf combination

### PR DESCRIPTION
**Pull request by tabVersion**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

as title, we disabled the combination by mistake. Now re-enabling it

related doc pr https://github.com/risingwavelabs/risingwave-docs/pull/2354

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [x] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

allow create table with `format upsert encode protobuf` but do note it as NOT RECOMMENDED. 

according to [docs.confluent.io/platform/7.6/control-center/topics/schema.html](https://docs.confluent.io/platform/7.6/control-center/topics/schema.html#c3-schemas-best-practices-key-value-pairs)

> Do not use Protobuf messages and JSON objects for key values. Avro does not guarantee deterministic serialization for maps or arrays, and Protobuf and JSON schema formats do not guarantee deterministic serialization for any object. Using these formats for key values will break topic partitioning.

Users may experience the upsert in wrong order.

For example, upsert_1 and upsert_2 occurs accordingly but goes into different partitions due to the partition key issue. upsert_2 may be ingested earlier than upsert_1 so the result in RisingWave can be wrong. 

</details>

### CODE REVIEW

**Comment by xiangjinwu** [comment]
> we disabled the combination by mistake

It was not by mistake. We never implemented parsing kafka key as protobuf until we decided not to parse and always `include key as` an additional `bytea` column.

**Comment by xiangjinwu** [comment]
Is it still draft? Any e2e test to make sure it works?

**Comment by tabVersion** [comment]
> > we disabled the combination by mistake
> 
> It was not by mistake. We never implemented parsing kafka key as protobuf until we decided not to parse and always `include key as` an additional `bytea` column.

IIRC, we the prev behavior is fetching schema from both key and value, comparing them, and inferring the pk (aligning with upsert avro). After migrating to the additional columns, the upsert protobuf is disabled.

**Comment by tabVersion** [comment]
> Is it still draft? Any e2e test to make sure it works?

I am working on e2e test.

**Comment by xiangjinwu** [comment]
> After migrating to the additional columns, the upsert protobuf is disabled.

Before the migration, `upsert protobuf` did not exist either 😂 The previous behavior of *fetching schema from both key and value* was only implemented for avro

**Review by xxchan — COMMENTED** [review]
Note that Flink also don't support Upsert Protobuf. https://nightlies.apache.org/flink/flink-docs-release-1.19/docs/connectors/table/formats/overview/

So this might be non-trivial

**Comment by xiangjinwu** [comment]
> Note that Flink also don't support Upsert Protobuf. https://nightlies.apache.org/flink/flink-docs-release-1.19/docs/connectors/table/formats/overview/
> 
> So this might be non-trivial

https://docs.confluent.io/platform/7.6/control-center/topics/schema.html#c3-schemas-best-practices-key-value-pairs
> Do not use Protobuf messages and JSON objects for key values. Avro does not guarantee deterministic serialization for maps or arrays, and Protobuf and JSON schema formats do not guarantee deterministic serialization for any object. Using these formats for key values will break topic partitioning.

**Comment by tabVersion** [comment]
>  Just my guess: [docs.confluent.io/platform/7.6/control-center/topics/schema.html](https://docs.confluent.io/platform/7.6/control-center/topics/schema.html#c3-schemas-best-practices-key-value-pairs)
> 
> > Do not use Protobuf messages and JSON objects for key values. Avro does not guarantee deterministic serialization for maps or arrays, and Protobuf and JSON schema formats do not guarantee deterministic serialization for any object. Using these formats for key values will break topic partitioning.

Agree, but the risk is not on our side. Maybe we can comment NOT RECOMMENDED on the doc.

File: e2e_test/source_inline/kafka/protobuf/basic.slt [hunk 1]
```diff
@@ -50,9 +66,16 @@ select min(id), max(id), max((sc).file_name) from sr_pb_test;
 ----
 0 19 source/context_019.proto
 
+query TT
+select min(rw_key), max(rw_key) from sr_pb_upsert;
+----
+\x7b226964223a20307d \x7b226964223a20397d
```
Comment (xiangjinwu) [inline e2e_test/source_inline/kafka/protobuf/basic.slt:72]: nit: `convert_from(rw_key, 'UTF-8')` to demonstrate how to see the key as readable json

File: src/frontend/src/handler/create_source.rs [hunk 2]
```diff
@@ -798,7 +798,10 @@ pub(crate) async fn bind_source_pk(
 
         // For all Upsert formats, we only accept one and only key column as primary key.
         // Additional KEY columns must be set in this case and must be primary key.
-        (Format::Upsert, encode @ Encode::Json | encode @ Encode::Avro) => {
+        (
+            Format::Upsert,
+            encode @ Encode::Json | encode @ Encode::Avro | encode @ Encode::Protobuf,
```
Comment (xiangjinwu) [inline src/frontend/src/handler/create_source.rs:803]: Given the comment above says *all Upsert formats*, shall we match `(Format::Upsert, _)` ?

**Review by xiangjinwu — APPROVED** [review]

File: src/frontend/src/handler/create_source.rs (hunk 2 above)
Comment (tabVersion) [reply src/frontend/src/handler/create_source.rs:803]: It makes sense for all encodes other than `Encode::Native` but we may proceed with caution. Let's add more encodes here as requested.

**Pull request merged**